# 1

The [Link](https://docs.google.com/spreadsheets/d/16i38oonuX1y1g7C_UAmiK9GkY7cS-64DfiDMNiR41LM/edit#gid=0) to the data set using to 2019 Winter Data Science Intern Challenge Data Set.

---

In [1]:
import pandas as pd
import numpy as np

In [138]:
#bringing in the dataset
df = pd.read_csv('2019 Winter Data Science Intern Challenge Data Set.csv')
df.head()

order_id  shop_id  user_id  order_amount  total_items payment_method  \
0         1       53      746           224            2           cash   
1         2       92      925            90            1           cash   
2         3       44      861           144            1           cash   
3         4       18      935           156            1    credit_card   
4         5       18      883           156            1    credit_card   

            created_at  
0  2017-03-13 12:36:56  
1  2017-03-03 17:38:52  
2   2017-03-14 4:23:56  
3  2017-03-26 12:43:37  
4   2017-03-01 4:35:11

In [76]:
#checking for missing values
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 7 columns):
order_id          5000 non-null int64
shop_id           5000 non-null int64
user_id           5000 non-null int64
order_amount      5000 non-null int64
total_items       5000 non-null int64
payment_method    5000 non-null object
created_at        5000 non-null object
dtypes: int64(5), object(2)
memory usage: 273.6+ KB


- No missing value

In [77]:
# confirm AOV given in question
AOV = df['order_amount'].mean()
AOV

3145.128

The figure is confirmed now lets check where the errors are coming from by looking at the `order_amount` column

In [67]:
df.sort_values(by='order_amount',ascending= False).head(50)

order_id  shop_id  user_id  order_amount  total_items payment_method  \
2153      2154       42      607        704000         2000    credit_card   
3332      3333       42      607        704000         2000    credit_card   
520        521       42      607        704000         2000    credit_card   
1602      1603       42      607        704000         2000    credit_card   
60          61       42      607        704000         2000    credit_card   
2835      2836       42      607        704000         2000    credit_card   
4646      4647       42      607        704000         2000    credit_card   
2297      2298       42      607        704000         2000    credit_card   
1436      1437       42      607        704000         2000    credit_card   
4882      4883       42      607        704000         2000    credit_card   
4056      4057       42      607        704000         2000    credit_card   
15          16       42      607        704000         2000    credit_card   
1104      1105       42      607        704000         2000    credit_card   
1562      1563       42      607        704000         2000    credit_card   
2969      2970       42      607        704000         2000    credit_card   
4868      4869       42      607        704000         2000    credit_card   
1362      1363       42      607        704000         2000    credit_card   
691        692       78      878        154350            6          debit   
2492      2493       78      834        102900            4          debit   
3724      3725       78      766         77175            3    credit_card   
4420      4421       78      969         77175            3          debit   
4192      4193       78      787         77175            3    credit_card   
3403      3404       78      928         77175            3          debit   
2690      2691       78      962         77175            3          debit   
2564      2565       78      915         77175            3          debit   
4715      4716       78      818         77175            3          debit   
1259      1260       78      775         77175            3    credit_card   
2906      2907       78      817         77175            3          debit   
3705      3706       78      828         51450            2    credit_card   
3101      3102       78      855         51450            2    credit_card   
4412      4413       78      756         51450            2          debit   
3167      3168       78      927         51450            2           cash   
490        491       78      936         51450            2          debit   
4079      4080       78      946         51450            2           cash   
1529      1530       78      810         51450            2           cash   
4311      4312       78      960         51450            2          debit   
2818      2819       78      869         51450            2          debit   
2821      2822       78      814         51450            2           cash   
617        618       78      760         51450            2           cash   
2512      2513       78      935         51450            2          debit   
511        512       78      967         51450            2           cash   
2452      2453       78      709         51450            2           cash   
493        494       78      983         51450            2           cash   
2495      2496       78      707         51450            2           cash   
4040      4041       78      852         25725            1           cash   
4918      4919       78      823         25725            1           cash   
1056      1057       78      800         25725            1          debit   
2922      2923       78      740         25725            1          debit   
2270      2271       78      855         25725            1    credit_card   
1193      1194       78      944         25725            1          debit   

               created_at  

### 1(a) Think about what could be going wrong with our calculation. Think about a better way to evaluate this data.

**Looking at the first few items on the dataset, the following could be observed.**

- There is an issues with the `order_amount` and `total_items` recorded for a couple of stores(42 and 78) and this affecting the AOV calulation


- The problematic transaction for shop_id 42 comes from user_id 607 which are all credit card transactions. The items in question have same `order_amount` and `total_items`,  this is most likely a system glitch.


- As regards the shop-78, the error comes only from the `order_amount`, which are absurdly high 


**A better way to calculate the AOV be to take one of the following approach**

- Isolate `stores 78 and 42`, perform a root cause analysis on the issues in these stores (determine the actual price per unit for each stores and fix issues - fill unit prices with average price across board).


- Calculate AOV, excluding stores 78 and 42

The first approach will be taken 

In [94]:
#adding the price per unit column

df['price_per_unit'] = df['order_amount']/df['total_items']

#re-arrange columns 
col = ['order_id', 'shop_id', 'user_id', 'order_amount', 'total_items','price_per_unit',
       'payment_method', 'created_at']

df = df[col]

- **Deep diving into both shops and  checking the price per unit**

In [95]:
df[df['shop_id'] == 42]

order_id  shop_id  user_id  order_amount  total_items  price_per_unit  \
15          16       42      607        704000         2000           352.0   
40          41       42      793           352            1           352.0   
60          61       42      607        704000         2000           352.0   
308        309       42      770           352            1           352.0   
409        410       42      904           704            2           352.0   
520        521       42      607        704000         2000           352.0   
834        835       42      792           352            1           352.0   
835        836       42      819           704            2           352.0   
938        939       42      808          1056            3           352.0   
979        980       42      744           352            1           352.0   
1104      1105       42      607        704000         2000           352.0   
1362      1363       42      607        704000         2000           352.0   
1364      1365       42      797          1760            5           352.0   
1367      1368       42      926          1408            4           352.0   
1436      1437       42      607        704000         2000           352.0   
1471      1472       42      907          1408            4           352.0   
1512      1513       42      946           352            1           352.0   
1520      1521       42      756           704            2           352.0   
1562      1563       42      607        704000         2000           352.0   
1602      1603       42      607        704000         2000           352.0   
1911      1912       42      739           704            2           352.0   
1929      1930       42      770           352            1           352.0   
2003      2004       42      934           704            2           352.0   
2018      2019       42      739           352            1           352.0   
2053      2054       42      951           352            1           352.0   
2153      2154       42      607        704000         2000           352.0   
2273      2274       42      747           704            2           352.0   
2297      2298       42      607        704000         2000           352.0   
2491      2492       42      868           704            2           352.0   
2609      2610       42      868           704            2           352.0   
2766      2767       42      970           704            2           352.0   
2835      2836       42      607        704000         2000           352.0   
2969      2970       42      607        704000         2000           352.0   
2987      2988       42      819          1056            3           352.0   
3332      3333       42      607        704000         2000           352.0   
3513      3514       42      726          1056            3           352.0   
3651      3652       42      830           352            1           352.0   
3697      3698       42      839           352            1           352.0   
3903      3904       42      975           352            1           352.0   
3998      3999       42      886           352            1           352.0   
4056      4057       42      607        704000         2000           352.0   
4231      4232       42      962           352            1           352.0   
4294      4295       42      859           704            2           352.0   
4326      4327       42      788           704            2           352.0   
4421      4422       42      736           704            2           352.0   
4625      4626       42      809           352            1           352.0   
4646      4647       42      607        704000         2000           352.0   
4745      4746       42      872           352            1           352.0   
4767      4768       42      720           704            2           352.0   
4868      4869       42      607        704000         2

- Even with the system error the price per unit is consistent at `$352`

The `order_amount` can be corrected to the right value `$704` and the `total_items` `which appears to be 2 instead of 2000`

In [100]:
df['order_amount'] == 704000

0       False
1       False
2       False
3       False
4       False
        ...  
4995    False
4996    False
4997    False
4998    False
4999    False
Name: order_amount, Length: 5000, dtype: bool

In [105]:
# correcting the order amount and total items
df.loc[df['order_amount'] == 704000,'order_amount'] = 704
df.loc[df['total_items'] == 2000,'total_items'] = 2

In [106]:
df[df['shop_id'] == 42]

order_id  shop_id  user_id  order_amount  total_items  price_per_unit  \
15          16       42      607           704            2           352.0   
40          41       42      793           352            1           352.0   
60          61       42      607           704            2           352.0   
308        309       42      770           352            1           352.0   
409        410       42      904           704            2           352.0   
520        521       42      607           704            2           352.0   
834        835       42      792           352            1           352.0   
835        836       42      819           704            2           352.0   
938        939       42      808          1056            3           352.0   
979        980       42      744           352            1           352.0   
1104      1105       42      607           704            2           352.0   
1362      1363       42      607           704            2           352.0   
1364      1365       42      797          1760            5           352.0   
1367      1368       42      926          1408            4           352.0   
1436      1437       42      607           704            2           352.0   
1471      1472       42      907          1408            4           352.0   
1512      1513       42      946           352            1           352.0   
1520      1521       42      756           704            2           352.0   
1562      1563       42      607           704            2           352.0   
1602      1603       42      607           704            2           352.0   
1911      1912       42      739           704            2           352.0   
1929      1930       42      770           352            1           352.0   
2003      2004       42      934           704            2           352.0   
2018      2019       42      739           352            1           352.0   
2053      2054       42      951           352            1           352.0   
2153      2154       42      607           704            2           352.0   
2273      2274       42      747           704            2           352.0   
2297      2298       42      607           704            2           352.0   
2491      2492       42      868           704            2           352.0   
2609      2610       42      868           704            2           352.0   
2766      2767       42      970           704            2           352.0   
2835      2836       42      607           704            2           352.0   
2969      2970       42      607           704            2           352.0   
2987      2988       42      819          1056            3           352.0   
3332      3333       42      607           704            2           352.0   
3513      3514       42      726          1056            3           352.0   
3651      3652       42      830           352            1           352.0   
3697      3698       42      839           352            1           352.0   
3903      3904       42      975           352            1           352.0   
3998      3999       42      886           352            1           352.0   
4056      4057       42      607           704            2           352.0   
4231      4232       42      962           352            1           352.0   
4294      4295       42      859           704            2           352.0   
4326      4327       42      788           704            2           352.0   
4421      4422       42      736           704            2           352.0   
4625      4626       42      809           352            1           352.0   
4646      4647       42      607           704            2           352.0   
4745      4746       42      872           352            1           352.0   
4767      4768       42      720           704            2           352.0   
4868      4869       42      607           704          

- Issues with shop 42 is now fixed

Next is shop 78

In [108]:
df[df['shop_id'] == 78]

order_id  shop_id  user_id  order_amount  total_items  price_per_unit  \
160        161       78      990         25725            1         25725.0   
490        491       78      936         51450            2         25725.0   
493        494       78      983         51450            2         25725.0   
511        512       78      967         51450            2         25725.0   
617        618       78      760         51450            2         25725.0   
691        692       78      878        154350            6         25725.0   
1056      1057       78      800         25725            1         25725.0   
1193      1194       78      944         25725            1         25725.0   
1204      1205       78      970         25725            1         25725.0   
1259      1260       78      775         77175            3         25725.0   
1384      1385       78      867         25725            1         25725.0   
1419      1420       78      912         25725            1         25725.0   
1452      1453       78      812         25725            1         25725.0   
1529      1530       78      810         51450            2         25725.0   
2270      2271       78      855         25725            1         25725.0   
2452      2453       78      709         51450            2         25725.0   
2492      2493       78      834        102900            4         25725.0   
2495      2496       78      707         51450            2         25725.0   
2512      2513       78      935         51450            2         25725.0   
2548      2549       78      861         25725            1         25725.0   
2564      2565       78      915         77175            3         25725.0   
2690      2691       78      962         77175            3         25725.0   
2773      2774       78      890         25725            1         25725.0   
2818      2819       78      869         51450            2         25725.0   
2821      2822       78      814         51450            2         25725.0   
2906      2907       78      817         77175            3         25725.0   
2922      2923       78      740         25725            1         25725.0   
3085      3086       78      910         25725            1         25725.0   
3101      3102       78      855         51450            2         25725.0   
3151      3152       78      745         25725            1         25725.0   
3167      3168       78      927         51450            2         25725.0   
3403      3404       78      928         77175            3         25725.0   
3440      3441       78      982         25725            1         25725.0   
3705      3706       78      828         51450            2         25725.0   
3724      3725       78      766         77175            3         25725.0   
3780      3781       78      889         25725            1         25725.0   
4040      4041       78      852         25725            1         25725.0   
4079      4080       78      946         51450            2         25725.0   
4192      4193       78      787         77175            3         25725.0   
4311      4312       78      960         51450            2         25725.0   
4412      4413       78      756         51450            2         25725.0   
4420      4421       78      969         77175            3         25725.0   
4505      4506       78      866         25725            1         25725.0   
4584      4585       78      997         25725            1         25725.0   
4715      4716       78      818         77175            3         25725.0   
4918      4919       78      823         25725            1         25725.0   

     payment_method           created_at  
160     credit_card   2017-03-12 5:56:57  
490           debit  2017-03-26 17:08:19  
493            cash  2017-03-16 21:39:35  
511            cash   2017-03-09 7:23:14  
617            cash  2017-03-18 11:18:42  
691           debit  2017-03-27 22

Appears there is little one can do about shop 78, without contacting the store or filling the `price_per_unit` is  with the `average price across other stores`

In [129]:
# calculating the median of the prices across board and fill unit price for shop 78
df.loc[df['shop_id'] == 78,'price_per_unit'] = df[(df['shop_id'] != 78)]['price_per_unit'].mean()
df[df['shop_id'] == 78]

order_id  shop_id  user_id  order_amount  total_items  price_per_unit  \
160        161       78      990    152.475575            1      152.475575   
490        491       78      936    304.951151            2      152.475575   
493        494       78      983    304.951151            2      152.475575   
511        512       78      967    304.951151            2      152.475575   
617        618       78      760    304.951151            2      152.475575   
691        692       78      878    914.853452            6      152.475575   
1056      1057       78      800    152.475575            1      152.475575   
1193      1194       78      944    152.475575            1      152.475575   
1204      1205       78      970    152.475575            1      152.475575   
1259      1260       78      775    457.426726            3      152.475575   
1384      1385       78      867    152.475575            1      152.475575   
1419      1420       78      912    152.475575            1      152.475575   
1452      1453       78      812    152.475575            1      152.475575   
1529      1530       78      810    304.951151            2      152.475575   
2270      2271       78      855    152.475575            1      152.475575   
2452      2453       78      709    304.951151            2      152.475575   
2492      2493       78      834    609.902301            4      152.475575   
2495      2496       78      707    304.951151            2      152.475575   
2512      2513       78      935    304.951151            2      152.475575   
2548      2549       78      861    152.475575            1      152.475575   
2564      2565       78      915    457.426726            3      152.475575   
2690      2691       78      962    457.426726            3      152.475575   
2773      2774       78      890    152.475575            1      152.475575   
2818      2819       78      869    304.951151            2      152.475575   
2821      2822       78      814    304.951151            2      152.475575   
2906      2907       78      817    457.426726            3      152.475575   
2922      2923       78      740    152.475575            1      152.475575   
3085      3086       78      910    152.475575            1      152.475575   
3101      3102       78      855    304.951151            2      152.475575   
3151      3152       78      745    152.475575            1      152.475575   
3167      3168       78      927    304.951151            2      152.475575   
3403      3404       78      928    457.426726            3      152.475575   
3440      3441       78      982    152.475575            1      152.475575   
3705      3706       78      828    304.951151            2      152.475575   
3724      3725       78      766    457.426726            3      152.475575   
3780      3781       78      889    152.475575            1      152.475575   
4040      4041       78      852    152.475575            1      152.475575   
4079      4080       78      946    304.951151            2      152.475575   
4192      4193       78      787    457.426726            3      152.475575   
4311      4312       78      960    304.951151            2      152.475575   
4412      4413       78      756    304.951151            2      152.475575   
4420      4421       78      969    457.426726            3      152.475575   
4505      4506       78      866    152.475575            1      152.475575   
4584      4585       78      997    152.475575            1      152.475575   
4715      4716       78      818    457.426726            3      152.475575   
4918      4919       78      823    152.475575            1      152.475575   

     payment_method           created_at  
160     credit_card   2017-03-12 5:56:57  
490           debit  2017-03-26 17:08:19  
493            cash  2017-03-16 21:39:35  
511            cash   2017-03-09 7:23:14  
617            cash  2017-03-18 11:18:42  
691           debit  2017-03-27 22

In [130]:
# correcting the order_amount

df.loc[df['shop_id'] == 78,
       'order_amount'] = df[(df['shop_id'] == 78)]['total_items'] * df[(df['shop_id'] != 78)]['price_per_unit'].mean()


df[df['shop_id'] == 78]

order_id  shop_id  user_id  order_amount  total_items  price_per_unit  \
160        161       78      990    152.475575            1      152.475575   
490        491       78      936    304.951151            2      152.475575   
493        494       78      983    304.951151            2      152.475575   
511        512       78      967    304.951151            2      152.475575   
617        618       78      760    304.951151            2      152.475575   
691        692       78      878    914.853452            6      152.475575   
1056      1057       78      800    152.475575            1      152.475575   
1193      1194       78      944    152.475575            1      152.475575   
1204      1205       78      970    152.475575            1      152.475575   
1259      1260       78      775    457.426726            3      152.475575   
1384      1385       78      867    152.475575            1      152.475575   
1419      1420       78      912    152.475575            1      152.475575   
1452      1453       78      812    152.475575            1      152.475575   
1529      1530       78      810    304.951151            2      152.475575   
2270      2271       78      855    152.475575            1      152.475575   
2452      2453       78      709    304.951151            2      152.475575   
2492      2493       78      834    609.902301            4      152.475575   
2495      2496       78      707    304.951151            2      152.475575   
2512      2513       78      935    304.951151            2      152.475575   
2548      2549       78      861    152.475575            1      152.475575   
2564      2565       78      915    457.426726            3      152.475575   
2690      2691       78      962    457.426726            3      152.475575   
2773      2774       78      890    152.475575            1      152.475575   
2818      2819       78      869    304.951151            2      152.475575   
2821      2822       78      814    304.951151            2      152.475575   
2906      2907       78      817    457.426726            3      152.475575   
2922      2923       78      740    152.475575            1      152.475575   
3085      3086       78      910    152.475575            1      152.475575   
3101      3102       78      855    304.951151            2      152.475575   
3151      3152       78      745    152.475575            1      152.475575   
3167      3168       78      927    304.951151            2      152.475575   
3403      3404       78      928    457.426726            3      152.475575   
3440      3441       78      982    152.475575            1      152.475575   
3705      3706       78      828    304.951151            2      152.475575   
3724      3725       78      766    457.426726            3      152.475575   
3780      3781       78      889    152.475575            1      152.475575   
4040      4041       78      852    152.475575            1      152.475575   
4079      4080       78      946    304.951151            2      152.475575   
4192      4193       78      787    457.426726            3      152.475575   
4311      4312       78      960    304.951151            2      152.475575   
4412      4413       78      756    304.951151            2      152.475575   
4420      4421       78      969    457.426726            3      152.475575   
4505      4506       78      866    152.475575            1      152.475575   
4584      4585       78      997    152.475575            1      152.475575   
4715      4716       78      818    457.426726            3      152.475575   
4918      4919       78      823    152.475575            1      152.475575   

     payment_method           created_at  
160     credit_card   2017-03-12 5:56:57  
490           debit  2017-03-26 17:08:19  
493            cash  2017-03-16 21:39:35  
511            cash   2017-03-09 7:23:14  
617            cash  2017-03-18 11:18:42  
691           debit  2017-03-27 22

Now shop-78 has been fixed

In [131]:
new_AOV = df['order_amount'].mean()
new_AOV

303.84517012515164

**The new AOV figure is `~$304`**

In [133]:
# calculating the AOV per store

df.groupby('shop_id')['order_amount'].mean().sort_values()

shop_id
92     162.857143
2      174.327273
32     189.976190
100    213.675000
53     214.117647
          ...    
81     384.000000
38     390.857143
90     403.224490
50     403.545455
42     669.490196
Name: order_amount, Length: 100, dtype: float64

---

### 1(b) What metric would you report for this dataset?

**Sales per employee** is a good metric that one can measure from this dataset since we have the `user_id` column and it is assumed that every employee has as `user_id`.

calculated by 
$$
\text{Sales per employee} = \frac{net sales}{number of employees}
$$


This metric is very helpful in making smart employment decisions, particularly when is comes to hiring and compensation. Setting a good KPI for this metric can help motivate employees to perform better

### 1(c) What is its value?

In [136]:
num_of_emp = df['user_id'].nunique()
total_sales = df['order_amount'].sum()

print(f'The total sales is ${round(total_sales,2)} and the total employee number is {num_of_emp}')

The total sales is $1519225.85 and the total employee number is 301


In [137]:
sale_per_employee = total_sales/num_of_emp

print(f'The sales per employee metric is ${round(sale_per_employee,2)}')

The sales per employee metric is $5047.26


---

**In Summary**

- The data was looked at to check for missing values


- The AOV was calculated and confirmed to be smilar to that of the question


- A deep dive was done on the data to determine where issues are coming from and it was discovered to be coming from `shop 42 and shop 78`


- Both shops above were critically looked; errors in shop 42 were corrected and the average price of stores were used to fill the unit price for shop 78 before issues were fixed.


- A new AOV was calculated using the new set of data


- `Sales per employee` was suggested as the very good metric that can help valuable insight about the data

---
# 2
The [Link](https://www.w3schools.com/SQL/TRYSQL.ASP?FILENAME=TRYSQL_SELECT_ALL) to the data set using to asnwer the questions

---

### 2(a) How many orders were shipped by Speedy Express in total?


- SQL code

`SELECT count(*) as num_of_orders
FROM Orders o
LEFT JOIN Shippers s on 
o.ShipperID = s.ShipperID
WHERE ShipperName = 'Speedy Express'
;`

**`54` orders were shipped by Speedy Express**

### 2(b) What is the last name of the employee with the most orders?

- SQL Code

`SELECT e.LastName
FROM
(
SELECT EmployeeID, count(*) as num_of_orders
FROM Orders
GROUP BY EmployeeID
ORDER BY num_of_orders DESC
LIMIT 1) as emp_o
LEFT JOIN Employees e 
on emp_o.EmployeeID = e.EmployeeID
;`


**The last name of the employee with the mist order is `Peacock` with `40 orders`**

### 2(c) What product was ordered the most by customers in Germany?


-- The problem will be approached by using CTEs; 1- ordersDetails which includes products, and 2- For Germany Orders


-- CTE 1 - order_details

`
WITH order_details AS
(
select *
from OrderDetails od
left Join Products p on 
od.ProductID = p.ProductID
)`
,


-- CTE 2 - Germany Order

`
ger_orders AS
(SELECT * 
FROM Orders o
Left JOIN Customers c on
o.CustomerID = c.CustomerID
where Country = 'Germany')
`

-- Bring it together and outputing results

`
select ProductName
from (
select os.ProductName, count(*) as num_of_orders
from ger_orders g 
left join order_details os
on g.OrderID = os.OrderID
group by os.ProductName
order by num_of_orders DESC
limit 1) as p_list;
`


**The most ordered product in Germany is `Gorgonzola Telino`**